### DISCUSSION TITLE
Imgur Authentication Fails to Redirect to Obsidian

### DISCUSSION THREAD

**Discussion by Alvin21Bon** [gavvvr/obsidian-imgur-plugin#48] [Getting help]
OS: Linux/GNU Debian Bookworm
DE: Gnome
Obsidian Version: 1.1.16
Browser: Firefox 102.10.0esr (with no adblocker of any kind)

I have been trying since yesterday to successfully authenticate, but every time it is unable to set the Obsidian plugin token, thus authentication is not successful even though "obsidian-imgur-plugin" shows as an application in my Imgur account. 

**Steps to reproduce**
1) Click authenticate in plugin settings
2) Type in Imgur credentials on the opened Firefox window
3) Press "Allow," with seemingly nothing happening the first time
4) Press "Allow" again which opens a new window displaying json data
`{"data":{"error":"Unauthorized","request":"\/oauth2\/authorize","method":"POST"},"success":false,"status":403}`

Afterwards, the plugin shows as a connected app in my Imgur settings, but the plugin still shows a "Not Authenticated" status.

**Comment by gavvvr** [comment]
Hi @Alvin21Bon 

I guess this is something related to your local OS setup and not a bug of a plugin. Based on your description I tend to think this can be a problem with your Obsidian installation which did not register a handler for `obsidian://...` URLs in your OS.

My OS is macOS, I just tried setting Firefox 112.0.1 as my default browser and tried to authenticate Imgur plugin in my Obsidian. It worked out without any problems. There was a single question from Firefox to open the redirected link with Obsidian. I pressed "Allow" and authentication has completed successfully.

I will convert this "issue" to "discussion". Hope you will find a solution for your problem soon. Please share the solution when you will find it, so it can be useful for other Linux users.

**Reply by Alvin21Bon** [reply]
Thank you for you reply, @gavvvr. Your suggestion helped me pinpoint what exactly was wrong. 

The problem was that the Obsidian URL handling was not set up in the `.desktop` file that I created for the Obsidian AppImage. Thus, typing in the command `xdg-open "obsidian://"` in my terminal yielded an error. 

# The Solution
1) First, make sure the dependencies are installed for URL handling
```
sudo apt install xdg-utils desktop-file-utils
```

2) Now, run the command `xdg-open "obsidian://"`, if this yields an error, then this could explain why the imgur-plugin fails to redirect to the Obsidian application

3) To solve this, first create an `Obsidian.desktop` file for obsidian with the following **example** content:
```
[Desktop Entry]
Version=1.1.16
Name=Obsidian
Exec=/usr/local/bin/obsidian %u   <-- IMPORTANT
Icon=/usr/share/pixmaps/obsidian.png
Terminal=false
Type=Application
StartupNotify=true
Categories=TextEditor;Markdown;NoteTaking;Office;
MimeType=text/html;x-scheme-handler/obsidian;    <-- IMPORTANT
```
**The entries labeled are the ones responsible for setting up URL handling. (%u in the `Exec` parameter makes sure URLS are passed to the AppImage file, and the `MimeType` parameter registers Obsidan to handle "obsidian://" URLS.)**

4) Now run the command `sudo update-desktop-database ~/.local/share/applications/`, change the directory to where you placed your `Obsidian.desktop` file. 

5) Now, run `xdg-open "obsidian://"` and confirm that Obsidian is now correctly being opened. 

...

After following these steps, I was able to successfully authenticate the plugin. Thank you! Hope this helps.
